### PR TITLE
#10

### PR DESCRIPTION
**Pull request by DMan113**
Знімок екрана (2)

### Description of Changes:

**1) Logging for the entire project:**

     _- A centralized logging system has been implemented across the project.
     - Logging is configured at the debug level, enabling detailed event tracking during application execution.
     - Log entries are stored in JSON format, making them easier to parse and integrate with monitoring and analysis tools._

**2) New permissions.py file:**

     _- A new DetailedPermissionLogging class has been added to handle logging of all events related to permission checks.
     - Logs now capture access attempts, successful and unsuccessful permission checks, including request details like IP address, method, and authentication status.
     - The InvestmentPermission class has been added, which checks user permissions based on their investment records, logging the outcome of each check._

**3) New utils/logging_utils.py file and users/permissions.py:**

     - Contains utility to convert to JSON  format
     - The log_event method has been added to standardize logging across different severity levels (debug, info, warning, error) and handle extra logging data._
     
     
     
### The default is DEBUG.
You can select a different level in the settings, changing it to INFO, WARNING, ERROR or CRITICAL.
photo_2024-10-25_05-45-13

### CODE REVIEW

**Review by uU-Uuu — APPROVED** [review]

**Review by VovaD199 — APPROVED** [review]

File: users/models.py [hunk 1]
```diff
@@ -66,22 +75,21 @@ def create_superuser(self, email, password=None, **extra_fields):
...
             User: Newly created superuser instance.
         """
+        logger.debug(f"Attempting to create superuser with email: {email}")
         extra_fields.setdefault('is_staff', True)
         extra_fields.setdefault('is_superuser', True)
         extra_fields.setdefault('is_active', True)
         user = self.create_user(email, password, **extra_fields)
-        logger.info(f"New superuser created: {email}")
+        logger.info("New superuser created", extra={'email': email})
         return user
```
[the remark this answers is not in the document]
Comment (DMan113) [reply users/models.py:85]: Got it! Thank you for your attention. Will check and update it.

File: users/permissions.py [hunk 2]
```diff
@@ -1 +1,159 @@
...
+from django.core.exceptions import ObjectDoesNotExist
+from investment_tracking.models import InvestmentTracking
 
+logger = logging.getLogger('users')
+
+
+class DetailedPermissionLogging(permissions.BasePermission):
+    def __init__(self):
+        self.log_prefix = self.__class__.__name__
+
```
[the remark this answers is not in the document]
Comment (DMan113) [reply users/permissions.py:14]: Done

File: users/models.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (DMan113) [reply users/models.py:85]: Added these checks

File: users/permissions.py [hunk 3]
```diff
@@ -1 +1,159 @@
...
+            'path': request.path,
+            'method': request.method,
+            'view': view.__class__.__name__,
+            'timestamp': timezone.now().isoformat(),
+            'user_id': getattr(request.user, 'id', None)
+        }
+
+    def log_event(self, level: str, message: str, extra_data: dict) -> None:
+        """Centralized event logging"""
+        try:
```
[the remark this answers is not in the document]
Comment (DMan113) [reply users/permissions.py:28]: Done

File: users/permissions.py [hunk 4]
```diff
@@ -1 +1,159 @@
...
+        # 1. Authentication check
+        if not request.user.is_authenticated:
+            self.log_event('warning', 'Unauthenticated access attempt', {
+                **base_log_data,
+                'reason': 'not_authenticated'
+            })
+            return False
+
+        # 2. Token check
+        auth_header = request.META.get('HTTP_AUTHORIZATION', '')
```
[the remark this answers is not in the document]
Comment (DMan113) [reply users/permissions.py:51]: Done

File: common/management/commands/clearcache.py [hunk 5]
```diff
@@ -1,10 +1,16 @@
+import logging
 from django.core.management.base import BaseCommand
 from django.core.cache import cache
 
+logger = logging.getLogger('django')
+
 
 class Command(BaseCommand):
     help = 'Clears the cache'
```
[the remark this answers is not in the document]
Comment (DMan113) [reply common/management/commands/clearcache.py]: Removed it

**Review by VovaD199 — APPROVED** [review]

**Pull request merged**